### PR TITLE
Dynamic Dashboard: Integrate new endpoint for product stock in Networking layer

### PR DESCRIPTION
**Pull request by itsmeichigo**
Part of #12748 

## Description

This PR updates the Networking layer to integrate the new endpoint `wc-analytics/reports/stock`. Changes include:
- Added a new model `ProductStock`.
- Added a mapper for the stock list.
- Updated `ProductsRemote` with a new method to fetch stock list.
- Added unit tests.

## Testing instructions

The endpoint hasn't been integrated yet so just CI passing is sufficient.

## Screenshots

N/A

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Networking/Networking/Mapper/ProductStockListMapper.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import Foundation
+
+/// Mapper: `[ProductStock]`
+///
+struct ProductStockListMapper: Mapper {
+    /// Site Identifier associated to the products that will be parsed.
+    ///
+    /// We're injecting this field via `JSONDecoder.userInfo` because SiteID is not returned in any of the Product Endpoints.
```
Comment (hafizrahman) [inline Networking/Networking/Mapper/ProductStockListMapper.swift:8]: 👍🏼 TIL about getting siteID this way.

File: Networking/Networking/Model/Product/ProductStock.swift [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        // some plugins alter the field value from String to Int or Decimal.
+        let sku = container.failsafeDecodeIfPresent(targetType: String.self,
+                                                    forKey: .sku,
+                                                    alternativeTypes: [.decimal(transform: { NSDecimalNumber(decimal: $0).stringValue })])
+
+        // Even though the API docs claim `manageStock` is a bool, it's possible that `"parent"`
+        // could be returned as well (typically with variations) — we need to account for this.
+        // A "parent" value means that stock mgmt is turned on + managed at the parent product-level, therefore
+        // we need to set this var as `true` in this situation.
+        // See: https://github.com/woocommerce/woocommerce-ios/issues/884 for more deets
```
Comment (hafizrahman) [inline Networking/Networking/Model/Product/ProductStock.swift:58]: Great research here as well.

File: Networking/Networking/Model/Product/ProductStock.swift [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        guard let siteID = decoder.userInfo[.siteID] as? Int64 else {
+            throw ProductDecodingError.missingSiteID
+        }
+
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        let productID = try container.decode(Int64.self, forKey: .productID)
+        let name = try container.decode(String.self, forKey: .name)
+
+        // Even though a plain install of WooCommerce Core provides String values,
+        // some plugins alter the field value from String to Int or Decimal.
```
Comment (hafizrahman) [inline Networking/Networking/Model/Product/ProductStock.swift:49]: Great research to find how plugins might alter this value.

File: Storage/Storage/Model/DashboardCard.swift [hunk 4]
```diff
@@ -6,14 +6,14 @@ public struct DashboardCard: Codable, Hashable, Equatable, GeneratedCopiable {
...
-    /// An available card will become invisible on the Dashboard, but stay visible on Customize, if `enabled` is set to false.
-    public let enabled: Bool
-
     /// The card's availability state for the site.
     /// To be set externally based on each card's availability check.
     public let availability: AvailabilityState
 
+    /// User-changeable setting in the Customize screen, whether to enable or disable an available card.
+    /// An available card will become invisible on the Dashboard, but stay visible on Customize, if `enabled` is set to false.
+    public let enabled: Bool
```
Comment (hafizrahman) [inline Storage/Storage/Model/DashboardCard.swift:15]: Thanks for doing this! I had included this is one of my PRs but let's get this one first instead.

**Review by hafizrahman — APPROVED** [review]
This looks good to me 👍🏼 

I think it will be useful to add a sample json and test to show how this deals with the case where plugins change the type for `sku`/`stockQuantity`, as well as the `parent` case for `manageStock`. But It's not a blocker for this PR.

**Comment by itsmeichigo** [comment]
Thanks @hafizrahman for the reviews!

> I think it will be useful to add a sample json and test to show how this deals with the case where plugins change the type for sku/stockQuantity, as well as the parent case for manageStock

Good point! I added the mock responses and tests for these cases in 32ceaec95e2fc6a0b5638f94c13ab2364e2d414a.

**Pull request merged**